### PR TITLE
Fix MAC Address endianness issue

### PR DESCRIPTION
**Pull request by komaljai**
"dstAddr" and "srcAddr" fields are of type "macaddr", which is stored in sk_buff in little endian form.
load_dword will convert big_endian to little_endian and load it to cpu. The issue is instead of calling load_dword (which does endianess conversion), we are calling __builtin_memcpy when there is annotation 'macaddr'

### CODE REVIEW

File: backends/tc/ebpfCodeGen.cpp [hunk 1]
```diff
@@ -407,14 +407,170 @@ void TCIngressPipelinePNA::emitLocalVariables(EBPF::CodeBuilder *builder) {
...
 
+void PnaStateTranslationVisitor::compileExtractField(const IR::Expression *expr,
+                                                  const IR::StructField *field, unsigned alignment,
+                                                  EBPF::EBPFType *type) {
+    unsigned widthToExtract = dynamic_cast<EBPF::IHasWidth *>(type)->widthInBits();
+    auto program = state->parser->program;
+    cstring msgStr;
+    cstring fieldName = field->name.name;
+
+    bool checkIfMAC = false;
```
Comment (fruffy) [inline backends/tc/ebpfCodeGen.cpp:433]: The eBPF back end has no knowledge of "MAC" . Rather it looks like you are trying to match on a specific bit-width?

File: backends/tc/ebpfCodeGen.cpp [hunk 2]
```diff
@@ -407,14 +407,170 @@ void TCIngressPipelinePNA::emitLocalVariables(EBPF::CodeBuilder *builder) {
...
+    }
+
+    builder->emitIndent();
+    builder->appendFormat("%s += %d", program->offsetVar.c_str(), widthToExtract);
+    builder->endOfStatement(true);
+
+    // eBPF can pass 64 bits of data as one argument passed in 64 bit register,
+    // so value of the field is printed only when it fits into that register
+    if (widthToExtract <= 64) {
+        cstring exprStr = expr->is<IR::PathExpression>()
```
Comment (fruffy) [inline backends/tc/ebpfCodeGen.cpp:552]: I believe you can use `toString` in both cases here. The outcome is the same.

File: backends/tc/ebpfCodeGen.cpp [hunk 3]
```diff
@@ -407,14 +407,170 @@ void TCIngressPipelinePNA::emitLocalVariables(EBPF::CodeBuilder *builder) {
...
+                              ? expr->to<IR::PathExpression>()->path->name.name
+                              : expr->toString();
+
+        if (expr->is<IR::Member>() && expr->to<IR::Member>()->expr->is<IR::PathExpression>() &&
+            isPointerVariable(
+                expr->to<IR::Member>()->expr->to<IR::PathExpression>()->path->name.name)) {
```
Comment (fruffy) [inline backends/tc/ebpfCodeGen.cpp:558]: Use `checkedTo` instead of `to.

**Review by fruffy — COMMENTED** [review]
This could fix #1122 but it looks like it only applies to the parser extraction right now. I am not quite sure about this implementation. What motivated this change?

Can this be implemented in a dedicated compiler pass?

File: backends/tc/ebpfCodeGen.cpp (hunk 1 above)
Comment (komaljai) [reply backends/tc/ebpfCodeGen.cpp:436]: This change is not in eBPF compiler, but in P4TC. If user gives annotation "macaddr", the expectation is not to convert the endianness. This is not on specific bit width, but annotation "macaddr".

File: backends/tc/ebpfCodeGen.cpp (hunk 2 above)
Comment (komaljai) [reply backends/tc/ebpfCodeGen.cpp:557]: Addressed.

File: backends/tc/ebpfCodeGen.cpp (hunk 3 above)
Comment (komaljai) [reply backends/tc/ebpfCodeGen.cpp]: Addressed.

File: backends/tc/ebpfCodeGen.cpp [hunk 4]
```diff
@@ -408,13 +408,175 @@ void TCIngressPipelinePNA::emitLocalVariables(EBPF::CodeBuilder *builder) {
...
+        cstring exprStr = expr->is<IR::PathExpression>()
+                              ? expr->to<IR::PathExpression>()->path->name.name
+                              : expr->toString();
+
+        if (expr->is<IR::Member>() &&
+            expr->checkedTo<IR::Member>()->expr->is<IR::PathExpression>() &&
+            isPointerVariable(
+                expr->to<IR::Member>()->expr->checkedTo<IR::PathExpression>()->path->name.name)) {
+            exprStr = exprStr.replace(".", "->");
+        }
```
Comment (fruffy) [inline backends/tc/ebpfCodeGen.cpp:566]:
```suggestion
        if (auto member = expr->to<IR>Member>()) {
            if (auto pathExpr = member->expr->to<IR::PathExpression()) {
                if (isPointerVariable(pathExpr->name.name)) {
                    exprStr = exprStr.replace(".", "->");
                }
            }
        }
```

File: backends/tc/ebpfCodeGen.cpp [hunk 5]
```diff
@@ -408,13 +408,175 @@ void TCIngressPipelinePNA::emitLocalVariables(EBPF::CodeBuilder *builder) {
...
+    builder->emitIndent();
+    builder->appendFormat("%s += %d", program->offsetVar.c_str(), widthToExtract);
+    builder->endOfStatement(true);
+
+    // eBPF can pass 64 bits of data as one argument passed in 64 bit register,
+    // so value of the field is printed only when it fits into that register
+    if (widthToExtract <= 64) {
+        cstring exprStr = expr->is<IR::PathExpression>()
+                              ? expr->to<IR::PathExpression>()->path->name.name
+                              : expr->toString();
```
Comment (fruffy) [inline backends/tc/ebpfCodeGen.cpp:559]:
```suggestion
        cstring exprStr = expr->toString();
```

File: backends/tc/ebpfCodeGen.cpp (hunk 4 above)
Comment (fruffy) [reply backends/tc/ebpfCodeGen.cpp:566]: In general, I would recommend checking early that `expr` is either a path expression or a member and then resolve variables as needed. Do some checking that it needs to be either of these values then you can avoid these casts.

File: testdata/p4tc_samples_outputs/global_action_example_02.c [hunk 6]
```diff
@@ -208,10 +208,10 @@ static __always_inline int process(struct __sk_buff *skb, struct my_ingress_head
                 goto reject;
             }
 
-            hdr->ethernet.dstAddr = (u64)((load_dword(pkt, BYTES(ebpf_packetOffsetInBits)) >> 16) & EBPF_MASK(u64, 48));
+            __builtin_memcpy(&hdr->ethernet.dstAddr, pkt + BYTES(ebpf_packetOffsetInBits), 6);
             ebpf_packetOffsetInBits += 48;
 
-            hdr->ethernet.srcAddr = (u64)((load_dword(pkt, BYTES(ebpf_packetOffsetInBits)) >> 16) & EBPF_MASK(u64, 48));
+            __builtin_memcpy(&hdr->ethernet.srcAddr, pkt + BYTES(ebpf_packetOffsetInBits), 6);
```
Comment (fruffy) [inline testdata/p4tc_samples_outputs/global_action_example_02.c:214]: Is this is a portable function? I believe clang requires language extensions to support this particular memcpy version.

Comment (fruffy) [reply testdata/p4tc_samples_outputs/global_action_example_02.c:214]: Would be nice to use a less compiler-specific version here.

**Review by usha1830 — APPROVED** [review]
LGTM

File: testdata/p4tc_samples_outputs/global_action_example_02.c (hunk 6 above)
Comment (komaljai) [reply testdata/p4tc_samples_outputs/global_action_example_02.c:214]: __builtin_memcpy is commonly used in eBPF code (not only P4-ebpf). There are many instances in same code where __builtin_memcpy is used.

File: backends/tc/ebpfCodeGen.cpp (hunk 4 above)
Comment (komaljai) [reply backends/tc/ebpfCodeGen.cpp]: Addressed

**Review by fruffy — APPROVED** [review]

**Comment by jafingerhut** [comment]
@komaljai Have you been able to get your status with the ONF CLA checker to pass yet?

**Comment by usha1830** [comment]
> @jafingerhut Not yet, I am yet to be added in CLA.

Merging this as the CLA check is passing now.

**Pull request merged**